Add FastAPI application with database integration and temperature man…

arsenmarkotskyi opened the pull request:
…agement features

- Created initial FastAPI application structure with routers for cities and temperatures.
- Implemented database models for City and Temperature using SQLAlchemy.
- Added error handling for HTTP exceptions and validation errors.
- Included async database operations with SQLite using aiosqlite.
- Defined schemas for city and temperature data validation with Pydantic.
- Integrated weather service to fetch current temperatures from an external API.
- Established requirements in requirements.txt for necessary packages.

arsenmarkotskyi committed:
Add FastAPI application with database integration and temperature management features

- Created initial FastAPI application structure with routers for cities and temperatures.
- Implemented database models for City and Temperature using SQLAlchemy.
- Added error handling for HTTP exceptions and validation errors.
- Included async database operations with SQLite using aiosqlite.
- Defined schemas for city and temperature data validation with Pydantic.
- Integrated weather service to fetch current temperatures from an external API.
- Established requirements in requirements.txt for necessary packages.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I am requesting changes because there are a few critical issues that will break functionality or violate explicit requirements. First, the city delete endpoint calls `await db.delete(db_city)` even though `AsyncSession.delete()` is synchronous; this will raise a `TypeError` at runtime and effectively break `DELETE /cities/{city_id}` (check around the delete handler in `app/routers/cities.py`). Second, your Pydantic response models for `City` and `Temperature` do not have `class Config: orm_mode = True`, but the routers return SQLAlchemy ORM objects with `response_model=...`; this will cause response validation/serialization errors for several endpoints (see `app/schemas/city.py` and `app/schemas/temperature.py`). Third, `CityUpdate` currently requires `name` because it inherits from `CityBase`, but your update logic treats fields as optional; this mismatch can prevent partial updates from working as intended and conflicts with the behavior of `PUT /cities/{city_id}`.

There are also a couple of correctness issues that can surface at runtime: `engine.dispose()` in `app/main.py` may not be awaitable depending on the SQLAlchemy version (verify and either remove `await` or call `engine.sync_engine.dispose()`), and the shutdown of the async engine should be tested so it doesn’t raise during app shutdown. In the temperatures router, database writes during `update_temperatures` are not wrapped in try/except with `await db.rollback()`, so a failing commit could leave the session in a bad state and break subsequent operations. Finally, the README and run instructions required by the task are missing, so the deliverables are incomplete even though the core code structure is solid. Once you fix the delete await bug, add `orm_mode`, adjust the update schema, and address the DB/shutdown/documentation issues, your solution will align well with the requirements and should be in great shape.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,60 @@
+from fastapi import FastAPI, Request
```
`Request` is imported but never used in this module — remove this import to avoid linter warnings and keep imports minimal.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,60 @@
+from fastapi import FastAPI, Request
+from fastapi.exceptions import RequestValidationError, HTTPException
```
Importing `HTTPException` from `fastapi.exceptions` is incorrect / fragile. Import `HTTPException` from `fastapi` (e.g. `from fastapi import HTTPException`) and keep `RequestValidationError` from `fastapi.exceptions` if you prefer that import. This avoids an ImportError on some FastAPI versions.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,60 @@
+    """
+    Creates database tables on application startup.
+    """
+    async with engine.begin() as conn:
+        await conn.run_sync(Base.metadata.create_all)
```
Good: you create DB tables at startup using the async engine and `run_sync` (this uses the sync `create_all` safely). No change required here — just confirming this matches the async SQLAlchemy setup in app/database.py.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,60 @@
+    """
+    Closes database connections on application shutdown.
+    """
+    await engine.dispose()
```
`await engine.dispose()` may be a problem depending on the SQLAlchemy version: `engine` is an AsyncEngine (see app/database.py). Verify whether `dispose()` is awaitable in your environment. If `dispose()` is synchronous, call `engine.dispose()` (no await) or call `engine.sync_engine.dispose()` to dispose the underlying sync engine. If your SQLAlchemy version provides an async `dispose()`, `await` is fine. Add a short comment or test to ensure shutdown doesn't raise at runtime.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,60 @@
+)
+
+# Register routers
+app.include_router(cities.router, prefix="/cities", tags=["cities"])
```
You register the routers for cities and temperatures here; ensure the routers themselves define paths relative to the prefixes (they do in app/routers/cities.py and app/routers/temperatures.py). This wiring satisfies the requirement to expose `/cities` and `/temperatures` endpoints.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,60 @@
+
+
+# Register error handlers
+app.add_exception_handler(HTTPException, http_exception_handler)
+app.add_exception_handler(RequestValidationError, validation_exception_handler)
+app.add_exception_handler(Exception, general_exception_handler)  
```
Exception handlers are registered centrally which is good. Double-check that the handler signatures in app/exceptions.py match the exception types (they do). Keep this centralized approach — it satisfies the requirement for graceful error handling.

mateacademy-ai-mentor commented on app/models/city.py:
```diff
@@ -0,0 +1,12 @@
+    __tablename__ = "cities"
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String, index=True)
```
Consider making `name` non-nullable and adding a length limit (e.g. `String(100)`) so the DB enforces data quality. This also helps with consistent Pydantic validation later. If city names must be unique, consider `unique=True` as well.

mateacademy-ai-mentor commented on app/models/city.py:
```diff
@@ -0,0 +1,12 @@
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String, index=True)
+    additional_info = Column(String, index=True)
```
Indexing `additional_info` is usually unnecessary and can bloat the index if this field stores arbitrary/long text. If you don't need to search by this field, remove `index=True`. Also consider allowing NULL (`nullable=True`) if it is optional.

mateacademy-ai-mentor commented on app/models/city.py:
```diff
@@ -0,0 +1,12 @@
+    name = Column(String, index=True)
+    additional_info = Column(String, index=True)
+
+    temperatures = relationship("Temperature", back_populates="city")
```
The relationship to `Temperature` should include cascade behavior (for example `cascade='all, delete-orphan'`) or you must ensure the `Temperature.city_id` FK is declared with `ondelete='CASCADE'`. Without either, deleting a `City` via the API may fail due to foreign-key constraints (this directly affects the `DELETE /cities/{city_id}` endpoint).

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,28 @@
+from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession, async_sessionmaker
```
The async imports look correct, but ensure you list and install the runtime dependency `aiosqlite` (and `SQLAlchemy>=2.0`) in your requirements so the URL `sqlite+aiosqlite://` works. This relates to checklist item #20 (SQLite + SQLAlchemy).

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,28 @@
+from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession, async_sessionmaker
+from sqlalchemy.orm import DeclarativeBase
```
`DeclarativeBase` is a SQLAlchemy 2.0 API. If you need to support older SQLAlchemy versions, you should either require SQLAlchemy>=2.0 in your README or fall back to `declarative_base()`; reference checklist item #20.

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,28 @@
+from sqlalchemy.orm import DeclarativeBase
+
+# URL for async SQLite (requires aiosqlite)
+SQLALCHEMY_DATABASE_URL = "sqlite+aiosqlite:///./cities.db"
```
Using a relative DB path (`./cities.db`) is fine for local development, but make it configurable (via env var or a settings file) and document this in the README so deployments have correct paths/permissions. This affects checklist items #7 and #28-#31 (README/run instructions).

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,28 @@
+)
+
+# Base class for models (SQLAlchemy 2.0 style)
+class Base(DeclarativeBase):
+    pass
```
You define `Base` (the model base) but this module does not create tables. To satisfy checklist items #7 and #14 (creating City and Temperature tables) ensure you run table creation at app startup — e.g. an async startup function that does `async with engine.begin(): await conn.run_sync(Base.metadata.create_all)` — or document where table creation happens.

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,28 @@
+
+
+# Dependency injection for getting database session
+async def get_db():
+    async with AsyncSessionLocal() as session:
+        yield session
```
`get_db` is implemented correctly as an async dependency (good for dependency injection per checklist item #23). Consider adding a return type hint like `-> AsyncGenerator[AsyncSession, None]` and a short docstring explaining its intended usage in route dependencies.

mateacademy-ai-mentor commented on app/exceptions.py:
```diff
@@ -0,0 +1,83 @@
+
+
+# Helper functions for creating standard errors
+def not_found_error(resource: str, resource_id: int = None) -> HTTPException:
```
The parameter `resource_id` is annotated as `int` but defaults to `None`. For correct typing and to avoid type-checker errors, consider using `Optional[int]` (from `typing`) or accept `Union[int, str]` if you expect non-integer IDs. This relates to robustness/type correctness for the helper (informational, checklist item #26).

mateacademy-ai-mentor replied:
Consider adding typing for the optional resource_id parameter (use `Optional[int]`) and import `Optional` from `typing`. This makes the function signature clearer about its optional parameter and aligns with type-hinting best practices.

mateacademy-ai-mentor replied:
Consider typing the optional resource_id parameter explicitly. Change the signature from `resource_id: int = None` to `resource_id: Optional[int] = None` and add `from typing import Optional` at the top. This makes the intent clearer and avoids using a bare `= None` with a non-optional type.

mateacademy-ai-mentor commented on app/exceptions.py:
```diff
@@ -0,0 +1,83 @@
+        content={
+            "detail": "Data validation error",
+            "errors": exc.errors(),
+            "body": exc.body
```
Accessing `exc.body` directly may raise an AttributeError in some contexts if the attribute is absent. Consider using `getattr(exc, "body", None)` or otherwise guarding this access so the validation handler never fails while preparing the error response (prevents cascading errors, checklist item #26).

mateacademy-ai-mentor replied:
Accessing `exc.body` may not be present on every RequestValidationError shape in all FastAPI/pydantic versions and could raise an AttributeError. Use a safe access pattern (for example `getattr(exc, 'body', None)`) when including it in the response to ensure the handler never crashes.

mateacademy-ai-mentor commented on app/routers/cities.py:
```diff
@@ -0,0 +1,96 @@
+from fastapi import APIRouter, Depends, HTTPException
```
`HTTPException` is imported but not used in this module. Remove it to keep imports minimal and avoid linter/warning noise.

mateacademy-ai-mentor commented on app/routers/cities.py:
```diff
@@ -0,0 +1,96 @@
+
+from app.database import get_db
+from app.models.city import City
+from app.schemas.city import CityCreate, CityUpdate, City as CitySchema
```
You import `CityCreate`, `CityUpdate`, and `City` schema here. Note: the current `CityUpdate` (in app/schemas/city.py) appears to inherit required fields from `CityBase`, so `name` is required for updates. Consider making update fields optional (or create a dedicated update schema with Optional fields) so partial updates work as intended (relates to checklist item #12).

mateacademy-ai-mentor commented on app/routers/cities.py:
```diff
@@ -0,0 +1,96 @@
+        raise bad_request_error(f"Error creating city: {str(e)}")
+
+
+@router.get("/", response_model=list[CitySchema])
```
When returning ORM model instances with `response_model=list[CitySchema]` or `response_model=CitySchema`, ensure the Pydantic schema has `Config.orm_mode = True` so FastAPI can serialize SQLAlchemy ORM objects. Without `orm_mode` you'll get serialization issues at runtime. (See app/schemas/city.py). This affects GET and POST responses (checklist items #8 and #9).

mateacademy-ai-mentor commented on app/routers/cities.py:
```diff
@@ -0,0 +1,96 @@
+        raise not_found_error("City", city_id)
+    
+    try:
+        await db.delete(db_city)
```
Bug: `await db.delete(db_city)` awaits a non-async method. `AsyncSession.delete()` is a regular (synchronous) method and should be called without `await`: `db.delete(db_city)`. Awaiting it will raise a TypeError at runtime — fix this to avoid a crash in DELETE /{city_id}.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,107 @@
+from datetime import datetime
+from typing import Optional
+
+from fastapi import APIRouter, Depends, HTTPException, Query
```
`HTTPException` is imported here but not used in this module. Remove the unused import to keep imports minimal and avoid linter warnings.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,107 @@
+from app.models.temperature import Temperature as TemperatureModel
+from app.schemas.temperature import Temperature
+from app.services.weather_service import get_temperature_for_city as fetch_temperature
+from app.exceptions import not_found_error, bad_request_error
```
`bad_request_error` is imported but never used in this file. Remove the unused import or use it where appropriate for consistent error handling.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,107 @@
+    - If `city_id` is specified: updates temperature only for the specified city
+    """
+    # Get cities from database
+    if city_id:
```
Use an explicit None check for `city_id`. Currently `if city_id:` treats falsy IDs (e.g. 0) as not-provided — prefer `if city_id is not None:` so the intent is clear and edge cases are handled correctly.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,107 @@
+    
+    # Fetch temperature for all cities in parallel
+    temperature_tasks = [fetch_temperature(city.name) for city in cities]
+    temperatures = await asyncio.gather(*temperature_tasks)
```
`asyncio.gather(*temperature_tasks)` will raise if any task raises an exception. Your weather service already returns None on errors, but to be more robust consider using `return_exceptions=True` and handling exceptions per-task, or wrap `gather` in try/except to ensure one failing external call doesn't cancel the whole update.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,107 @@
+                "error": "Failed to fetch temperature"
+            })
+    
+    await db.commit()
```
Database writes (adding temperature objects and committing) are not wrapped in error handling. If `db.commit()` fails an exception will propagate and you currently do not call `await db.rollback()`. Wrap the DB operations in try/except and rollback on failure (similar pattern used in app/routers/cities.py) to meet the requirement for graceful error handling.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,107 @@
+    }
+
+
+@router.get("/", response_model=list[Temperature])
```
The endpoint `get_temperatures` uses `response_model=list[Temperature]`. Ensure the Pydantic `Temperature` schema supports ORM objects (e.g. has `class Config: orm_mode = True`) so FastAPI can serialize SQLAlchemy model instances returned by the query. If `orm_mode` is missing the response serialization may fail.

mateacademy-ai-mentor commented on app/schemas/city.py:
```diff
@@ -0,0 +1,16 @@
+class CityCreate(CityBase):
+    pass
+
+class CityUpdate(CityBase):
```
`CityUpdate` currently inherits required fields from `CityBase`, so `name` will be required. The update router expects optional fields (it checks `if city.name is not None:`). Make fields optional here (e.g. `name: Optional[str] = None`, `additional_info: Optional[str] = None`) or define `CityUpdate` separately with Optional fields to allow partial updates. This addresses checklist item #12 and matches app/routers/cities.py.

mateacademy-ai-mentor commented on app/schemas/city.py:
```diff
@@ -0,0 +1,16 @@
+class CityUpdate(CityBase):
+    pass
+
+class City(CityBase):
+    id: int
```
Add Pydantic ORM configuration so response models can be created from SQLAlchemy ORM objects. For example add:

class Config:
    orm_mode = True

Either add this `Config` to `CityBase` (so all derived models inherit it) or to the `City` response model. Without `orm_mode = True`, FastAPI will have trouble serializing SQLAlchemy model instances returned by the routers. This is required to satisfy usage with response_model in the routers.

mateacademy-ai-mentor commented on app/schemas/temperature.py:
```diff
@@ -0,0 +1,13 @@
+class TemperatureCreate(TemperatureBase):
+    pass
+
+class Temperature(TemperatureBase):
+    id: int
+    date_time: datetime
```
Add a Pydantic `Config` with `orm_mode = True` to the response model so FastAPI can serialize SQLAlchemy ORM objects returned by the routers. Without `orm_mode` returning ORM instances as `response_model` will raise validation/serialization errors at runtime (relates to checklist items #16/#17/#19). Example: `class Config: orm_mode = True`.

mateacademy-ai-mentor commented on app/schemas/temperature.py:
```diff
@@ -0,0 +1,13 @@
+    city_id: int
+    temperature: float
+
+class TemperatureCreate(TemperatureBase):
+    pass
```
`TemperatureCreate` currently inherits only `city_id` and `temperature`, which is correct for creation. Ensure you intentionally omit `date_time` and `id` from the create model since these are set by the DB — this matches checklist item #13.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,44 @@
+    try:
+        async with httpx.AsyncClient(timeout=10.0) as client:
+            # URL for wttr.in API (free, no API key required)
+            url = f"https://wttr.in/{city_name}?format=j1"
```
When building the URL, `city_name` may contain spaces or special characters which can produce invalid URLs. Consider URL-encoding the city name (for example using `urllib.parse.quote_plus(city_name)`) before interpolating into the URL so requests work for multi-word city names.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,44 @@
+            response = await client.get(url)
+            response.raise_for_status()  # Raises exception for HTTP errors
+            
+            data = response.json()
```
`response.json()` is a synchronous method on httpx Response but is okay for typical small JSON payloads. If you expect many/large responses or want non-blocking parsing, note this and consider alternatives; otherwise this is acceptable here.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,44 @@
+        httpx.HTTPError: If HTTP request error occurred
+    """
+    try:
+        async with httpx.AsyncClient(timeout=10.0) as client:
```
Creating a new `AsyncClient` for every call is correct and safe but can be inefficient when updating many cities. For better performance when fetching temperatures in parallel, consider reusing a single `AsyncClient` instance (e.g., a module-level client or dependency) to avoid repeated TCP/TLS handshakes.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,44 @@
+            data = response.json()
+            
+            # Extract temperature from response
+            temperature = float(data["current_condition"][0]["temp_C"])
```
Parsing `data["current_condition"][0]["temp_C"]` is fine and you already catch `KeyError/IndexError/ValueError`. If you want clearer error messages, consider validating the expected structure before casting to `float`. Current approach is acceptable since parsing errors are handled and logged.

arsenmarkotskyi committed:
Enhance database and API functionality

- Updated database models to include constraints and relationships for better data integrity.
- Improved application shutdown process to close HTTP client and database connections.
- Refactored temperature update logic to handle exceptions and ensure robust error logging.
- Introduced a reusable HTTP client for improved performance in fetching temperatures.
- Updated Pydantic schemas to support partial updates and added configuration for attribute mapping.

mateacademy-ai-mentor commented on app/models/city.py:
```diff
@@ -0,0 +1,16 @@
+    __tablename__ = "cities"
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String, index=True, max_length=100)
```
`String` length is passed via an unsupported keyword `max_length` here. SQLAlchemy expects the length as a positional/constructor argument (e.g. `String(100)`) — change to: `name = Column(String(100), index=True)` to ensure the column length is applied correctly and to avoid runtime errors.

mateacademy-ai-mentor commented on app/models/city.py:
```diff
@@ -0,0 +1,16 @@
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String, index=True, max_length=100)
+    additional_info = Column(String, nullable=True)
```
Optional: `additional_info` has no length/type specified. If you expect long arbitrary text, consider using `Text` or `String(<length>)` to make the intent explicit (not strictly required but improves schema clarity).

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,36 @@
+# DeclarativeBase is the modern SQLAlchemy 2.0 API
+# For SQLAlchemy < 2.0, use: from sqlalchemy.ext.declarative import declarative_base
+#
+# Note: Database tables (City, Temperature) are created automatically on application startup
+# See app/main.py startup() event handler: Base.metadata.create_all()
```
The comment on lines 27-28 states `Base.metadata.create_all()` is called in `app/main.py`. With an async engine you must create tables via a synchronous call run through the async engine, e.g. `async with engine.begin() as conn: await conn.run_sync(Base.metadata.create_all)`. Ensure your startup handler uses `run_sync` or the engine's `sync_engine` so table creation works at runtime. This is relevant to table creation for the Temperature/City models (checklist items #7 and #14).

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,36 @@
+
+
+# Dependency injection for getting database session
+async def get_db():
+    async with AsyncSessionLocal() as session:
+        yield session
```
Consider adding an explicit return/annotation for the dependency: `async def get_db() -> AsyncGenerator[AsyncSession, None]:`. This makes the dependency type clear to readers and linters and aligns with dependency injection best practices (checklist item #24).

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,65 @@
+    Note: In SQLAlchemy 2.0+, AsyncEngine.dispose() is awaitable.
+    For older versions, use engine.sync_engine.dispose() instead.
+    """
+    await engine.dispose()
```
Calling `await engine.dispose()` unconditionally can raise if `AsyncEngine.dispose()` is not awaitable in the SQLAlchemy version used. Per checklist item #31, either call `engine.sync_engine.dispose()` for sync disposal or detect/handle both cases. Consider a safe approach such as trying `await engine.dispose()` and falling back to `engine.sync_engine.dispose()` on TypeError, or explicitly calling `engine.sync_engine.dispose()` when using SQLAlchemy < 2.0.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,65 @@
+    For older versions, use engine.sync_engine.dispose() instead.
+    """
+    await engine.dispose()
+    await close_http_client()
```
Ensure `close_http_client()` is an async function and that failures during shutdown don't propagate and crash the app. Wrap shutdown steps in try/except and log errors so a failing client close doesn't prevent a clean shutdown (checklist items #31 and #32).

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,65 @@
+    """
+    Creates database tables on application startup.
+    """
+    async with engine.begin() as conn:
+        await conn.run_sync(Base.metadata.create_all)
```
Startup code uses `await conn.run_sync(Base.metadata.create_all)`, which is the correct pattern for creating tables with an AsyncEngine; ensure `Base` is the declarative metadata that includes both `City` and `Temperature` models so the DB tables are created as required by checklist items #6 and #7.

mateacademy-ai-mentor commented on app/exceptions.py:
```diff
@@ -0,0 +1,83 @@
+logger = logging.getLogger(__name__)
+
+
+async def http_exception_handler(request: Request, exc: HTTPException) -> JSONResponse:
```
These handler functions look correct, but remember to register them with the FastAPI application (for example: `app.add_exception_handler(HTTPException, http_exception_handler)` etc.). Also avoid returning detailed internal errors to clients — ensure registration/site code doesn't expose sensitive internals.

mateacademy-ai-mentor commented on app/exceptions.py:
```diff
@@ -0,0 +1,83 @@
+    """
+    Handler for data validation errors (422 Unprocessable Entity).
+    """
+    logger.warning(f"Validation error: {exc.errors()}")
```
Be cautious about logging `exc.errors()` at warning level — validation error payloads can be verbose and potentially contain sensitive user data. Consider logging a summarized message at warning and the full `exc.errors()` at `debug` level, or sanitize before logging.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,129 @@
+    Gets a list of all temperature records.
+    Can be filtered by city_id through query parameter.
+    """
+    if city_id:
```
In `get_temperatures` you check `if city_id:` which treats falsy values (e.g. 0) as not provided. For consistency and correctness use `if city_id is not None:` (same pattern used in `update_temperatures`) to avoid accidentally ignoring valid IDs.

mateacademy-ai-mentor commented on app/models/temperature.py:
```diff
@@ -0,0 +1,16 @@
+    __tablename__ = "temperatures"
+
+    id = Column(Integer, primary_key=True, index=True)
+    city_id = Column(Integer, ForeignKey("cities.id", ondelete="CASCADE"))
```
Make the foreign key column explicitly non-nullable to enforce that every Temperature must reference a City and to match the Pydantic schema where `city_id` is required. For example: `city_id = Column(Integer, ForeignKey("cities.id", ondelete="CASCADE"), nullable=False)`.

mateacademy-ai-mentor commented on app/models/temperature.py:
```diff
@@ -0,0 +1,16 @@
+
+    id = Column(Integer, primary_key=True, index=True)
+    city_id = Column(Integer, ForeignKey("cities.id", ondelete="CASCADE"))
+    date_time = Column(DateTime, default=datetime.utcnow)
```
Consider whether you need timezone-aware timestamps or a database-level default. Using `default=datetime.utcnow` is fine for application-side defaults, but if you expect timezone-aware datetimes prefer `datetime.now(timezone.utc)` or store UTC explicitly; alternatively, use a server default if you want the DB to populate the timestamp.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,70 @@
+
+# Reusable HTTP client for better performance when making multiple parallel requests
+# This avoids repeated TCP/TLS handshakes
+_http_client: Optional[httpx.AsyncClient] = None
```
The module-level `_http_client` is lazily initialized. If two coroutines call `get_http_client()` at the same time you can end up creating multiple AsyncClient instances. Consider creating the client at application startup (and storing it in this module) or protect creation with an `asyncio.Lock` to avoid the race.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,70 @@
+_http_client: Optional[httpx.AsyncClient] = None
+
+
+def get_http_client() -> httpx.AsyncClient:
+    """
+    Returns a singleton AsyncClient instance for HTTP requests.
+    Creates the client on first call if it doesn't exist.
+    """
+    global _http_client
+    if _http_client is None:
+        _http_client = httpx.AsyncClient(timeout=10.0)
```
`get_http_client()` constructs an `httpx.AsyncClient` synchronously which works, but creating the client at startup (FastAPI `@app.on_event('startup')`) and returning it here is more robust and avoids the lazy-init race described above.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,70 @@
+        # Use reusable HTTP client for better performance
+        client = get_http_client()
+        # URL for wttr.in API (free, no API key required)
+        url = f"https://wttr.in/{city_name}?format=j1"
```
When building the wttr.in URL `f"https://wttr.in/{city_name}?format=j1"` you should URL-encode `city_name` (e.g. `urllib.parse.quote_plus`) to handle spaces and special characters which can otherwise produce invalid requests.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,70 @@
+        response = await client.get(url)
+        response.raise_for_status()  # Raises exception for HTTP errors
+        
+        data = response.json()
```
`response.json()` is synchronous and okay for typical payload sizes, but be aware it performs JSON parsing in the current thread. For very large payloads this could block — usually acceptable here, but keep in mind for higher-load scenarios.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,70 @@
+        logger.info(f"Fetched temperature for {city_name}: {temperature}°C")
+        return temperature
+            
+    except httpx.HTTPError as e:
```
Catching `httpx.HTTPError` will handle many common httpx exceptions, but you may want to explicitly handle or log `httpx.RequestError` and `httpx.HTTPStatusError` for clearer diagnostics (or confirm the base class used here covers the specific versions in your requirements).

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

I am requesting changes because there is at least one critical, functionality‑breaking issue that will prevent the app from working as intended. The `City` model defines `name = Column(String, max_length=100, index=True)` (in `app/models/city.py`), but SQLAlchemy does not support `max_length` as a keyword argument; this will raise at import/runtime and must be changed to something like `name = Column(String(100), index=True)` so the application can start correctly. In addition, the repository is missing the required `README.md` with run instructions, design choices, and assumptions, which is part of the explicit deliverables and means the submission is not complete from a user/setup perspective. 

The other issues are smaller but worth addressing when you revise: in `app/main.py` the unconditional `await engine.dispose()` can fail depending on the SQLAlchemy version, so use a version‑safe pattern (for example, call `engine.sync_engine.dispose()` or guard the `await` with a fallback) and wrap shutdown steps (DB dispose, `close_http_client()`) in `try/except` so shutdown errors don’t crash the app. Also consider tightening a few best practices: add a concrete return type to `get_db` (`AsyncGenerator[AsyncSession, None]`), change `if city_id:` to `if city_id is not None:` in `get_temperatures`, make `Temperature.city_id` non‑nullable with `ondelete="CASCADE"`, and URL‑encode `city_name` in the weather service. Overall, the core architecture and async patterns look solid; once you fix the City model bug and add the README (plus, ideally, the shutdown/cleanup tweaks), your solution will fully meet the requirements.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

arsenmarkotskyi committed:
Refactor application shutdown and enhance error handling

- Introduced logging for application shutdown processes to capture errors without crashing.
- Updated database model constraints for City and Temperature to ensure data integrity.
- Improved temperature retrieval logic by URL-encoding city names to handle special characters.
- Enhanced query parameter handling in temperature retrieval to check for None values.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -0,0 +1,83 @@
+    For older versions, use engine.sync_engine.dispose() instead.
+    """
+    # Close database connections
+    try:
+        await engine.dispose()
+        logger.info("Database connections closed successfully")
+    except Exception as e:
+        logger.error(f"Error closing database connections: {e}", exc_info=True)
```
Shutdown currently attempts `await engine.dispose()` but does not implement a version-safe fallback. For compatibility with older SQLAlchemy versions (where `dispose()` is synchronous) consider trying `await engine.dispose()` and, if that raises AttributeError/TypeError, call `engine.sync_engine.dispose()` instead; keep the outer try/except to log any other unexpected errors. This prevents shutdown failures across SQLAlchemy versions and follows the previous review guidance.

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,36 @@
+from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession, async_sessionmaker
```
Add an explicit import for the async generator return type used by the dependency. For example: `from typing import AsyncGenerator`. This import should be added near the top with the other imports so you can annotate `get_db`'s return type.

mateacademy-ai-mentor commented on app/database.py:
```diff
@@ -0,0 +1,36 @@
+
+
+# Dependency injection for getting database session
+async def get_db():
```
Annotate the `get_db` dependency with a concrete return type to make the dependency contract explicit and satisfy the checklist. For example: `async def get_db() -> AsyncGenerator[AsyncSession, None]:`. This clarifies the function yields an `AsyncSession` and improves type-checking and readability.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,129 @@
+    
+    # Fetch temperature for all cities in parallel
+    # Use return_exceptions=True to ensure one failing call doesn't cancel the whole update
+    temperature_tasks = [fetch_temperature(city.name) for city in cities]
+    temperature_results = await asyncio.gather(*temperature_tasks, return_exceptions=True)
```
When creating many concurrent requests you may hit external API rate limits or exhaust resources. Consider throttling concurrency (for example, use an asyncio.Semaphore or a bounded gather approach) instead of launching all coroutines at once with a plain list comprehension. This applies to the temperature task creation and gather call.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,129 @@
+            db_temperature = TemperatureModel(
+                city_id=city.id,
+                temperature=temperature,
+                date_time=datetime.utcnow()
```
You're storing `date_time=datetime.utcnow()` which is a naive UTC datetime. Prefer a timezone-aware timestamp (for example `datetime.now(timezone.utc)`) or explicitly document that stored datetimes are naive-UTC. This avoids ambiguity when the app/client expects timezone-aware values.

mateacademy-ai-mentor commented on app/routers/temperatures.py:
```diff
@@ -0,0 +1,129 @@
+        result = await db.execute(select(City))
+        cities = result.scalars().all()
+        
+        if not cities:
+            raise not_found_error("Cities")
```
When no cities exist you raise `not_found_error("Cities")`. Consider using a singular resource name or a more descriptive message (e.g., `No cities registered`) so the API consumer gets a clearer message. The current message may be ambiguous.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,73 @@
+_http_client: Optional[httpx.AsyncClient] = None
+
+
+def get_http_client() -> httpx.AsyncClient:
```
Consider adding an explicit return type to `get_http_client()` and making client initialization safer. Concurrent coroutines could race when `_http_client` is None and both try to create a client — either initialize the client at application startup (recommended) or guard creation with an asyncio.Lock to ensure a single instance is created.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,73 @@
+    return _http_client
+
+
+async def close_http_client():
+    """
+    Closes the HTTP client. Should be called on application shutdown.
+    """
+    global _http_client
+    if _http_client is not None:
+        await _http_client.aclose()
```
Wrap the client close in a try/except and add a return type (`-> None`). If `_http_client.aclose()` raises for any reason it will propagate to callers; since shutdown logic already wraps calls, handling errors here (and logging them) makes the function more self-contained and robust.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,73 @@
+        client = get_http_client()
+        # URL for wttr.in API (free, no API key required)
+        # URL-encode city name to handle spaces and special characters
+        encoded_city_name = quote_plus(city_name)
```
Good: city names are URL-encoded with `quote_plus` before calling wttr.in, which satisfies the requirement to URL-encode `city_name` for the external API.

mateacademy-ai-mentor commented on app/services/weather_service.py:
```diff
@@ -0,0 +1,73 @@
+        response = await client.get(url)
+        response.raise_for_status()  # Raises exception for HTTP errors
+        
+        data = response.json()
```
Minor: `response.json()` is synchronous but acceptable here (small payload). If you expect large payloads or want to be explicit, you can parse `response.text` or use `response.json()` as is — current usage is fine and handled by the existing try/except blocks that catch parsing errors.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

I am approving your solution as an **educational override** so you can keep progressing while you address a few remaining polish items.

You’ve done a strong job overall: the FastAPI app is well-structured, async DB access is correctly wired, the City CRUD and temperature endpoints behave as required, and the weather service correctly URL-encodes city names and reuses an async client. The main functional gap is the missing `README.md` with run instructions and design notes, which is a required deliverable; adding a short README will satisfy that requirement and help others run your app. A few smaller improvements will make the code more robust and type-safe: in `app/database.py` add a concrete return type like `AsyncGenerator[AsyncSession, None]` to `get_db`, in `app/main.py` make shutdown compatible with older SQLAlchemy by falling back to `engine.sync_engine.dispose()` if `await engine.dispose()` isn’t supported, and in `app/exceptions.py` use `Optional[int]` for `resource_id` and access `exc.body` with `getattr(exc, "body", None)` to avoid rare crashes. You can also consider limiting concurrency in `app/routers/temperatures.py` (e.g., via `asyncio.Semaphore`) when updating many cities, and tightening the http client lifecycle and typing in `app/services/weather_service.py` (explicit return types, guarded single initialization, safe `aclose()` with try/except).

Overall, the key requirements are met and your structure is clean; addressing these small issues will bring your project to a very solid, production-ready level.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
